README.md
# Jab

Jab is a simple hierarchical dependency injection framework for Flutter. It offers easy dependency overriding in the Widget tree and provides out-of-the-box support for the BLoC pattern.

CHANGELOG.md
## [0.0.1] - 5 May 2020

-   Initial Jab library implementation

## [0.0.2] - 5 May 2020

-   Added a usage example

## [0.0.3] - 5 May 2020

-   Added the option to use root as the main dependency injector (e.g. for testing purposes).

## [0.0.4] - 2 June 2020

-   Breaking change: If an instance of Service cannot be found in the Widget tree, Jab will attempt to find the Service instance in all injectors which are currently initialized.

## [0.0.5] - 6 February 2020

-   Complete Architecture redesign with focus on injector modularity and null safety support.


jab/test/jab_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:jab/jab.dart';

/// [ServiceA] has no dependencies
class ServiceA {}

/// [ServiceB] depends on [ServiceA]
class ServiceB {
  ServiceB(this.serviceA);

  final ServiceA serviceA;
}

/// [ServiceC] depends on both [ServiceA] and [ServiceB]
class ServiceC {
  ServiceC(this.serviceA, this.serviceB);

  final ServiceA serviceA;
  final ServiceB serviceB;
}

/// [ServiceD] depends on [ServiceC]
class ServiceD {
  ServiceD(this.serviceC);

  final ServiceC serviceC;
}

final providers = <JabFactory>[
  (_) => ServiceA(),
  (jab) => ServiceB(jab()),
  (jab) => ServiceC(jab(), jab()),
  (jab) => ServiceD(jab()),
];

void main() {
  group('JabInjector', () {
    tearDown(() {
      JabInjector.root.clear();
    });

    test('get - ServiceA', () {
      Jab.provideForRoot(providers);
      final serviceA = JabInjector.root.get<ServiceA>();
      expect(serviceA, isA<ServiceA>());
    });

    test('get - ServiceB', () {
      Jab.provideForRoot(providers);
      final serviceB = JabInjector.root.get<ServiceB>();
      expect(serviceB, isA<ServiceB>());
      expect(serviceB.serviceA, isA<ServiceA>());
    });

    test('get - ServiceC', () {
      Jab.provideForRoot(providers);
      final serviceC = JabInjector.root.get<ServiceC>();
      expect(serviceC, isA<ServiceC>());
      expect(serviceC.serviceA, isA<ServiceA>());
      expect(serviceC.serviceB, isA<ServiceB>());
    });

    test('get - all in order', () {
      Jab.provideForRoot(providers);

      final serviceA = JabInjector.root.get<ServiceA>();
      expect(serviceA, isA<ServiceA>());

      final serviceB = JabInjector.root.get<ServiceB>();
      expect(serviceB, isA<ServiceB>());
      expect(serviceB.serviceA, serviceA);

      final serviceC = JabInjector.root.get<ServiceC>();
      expect(serviceC, isA<ServiceC>());
      expect(serviceC.serviceA, serviceA);
      expect(serviceC.serviceB, serviceB);

      final serviceD = JabInjector.root.get<ServiceD>();
      expect(serviceD, isA<ServiceD>());
      expect(serviceD.serviceC, serviceC);
    });

    test('get - all reversed', () {
      Jab.provideForRoot(providers);

      // Note: ServiceD creates all Services.

      final serviceD = JabInjector.root.get<ServiceD>();
      expect(serviceD, isA<ServiceD>());
      expect(serviceD.serviceC, isA<ServiceC>());

      final serviceC = JabInjector.root.get<ServiceC>();
      expect(serviceC, serviceD.serviceC);

      final serviceB = JabInjector.root.get<ServiceB>();
      expect(serviceB, serviceD.serviceC.serviceB);

      final serviceA = JabInjector.root.get<ServiceA>();
      expect(serviceA, serviceD.serviceC.serviceA);
      expect(serviceA, serviceD.serviceC.serviceB.serviceA);
    });

    test('get - onCreate callback in order', () {
      final services = [];
      Jab.setOnCreate(services.add);

      Jab.provideForRoot(providers);

      final serviceA = JabInjector.root.get<ServiceA>();
      expect(services.length, 1);
      expect(services, containsAllInOrder([serviceA]));

      final serviceB = JabInjector.root.get<ServiceB>();
      expect(services.length, 2);
      expect(services, containsAllInOrder([serviceA, serviceB]));

      final serviceC = JabInjector.root.get<ServiceC>();
      expect(services.length, 3);
      expect(services, containsAllInOrder([serviceA, serviceB, serviceC]));

      final serviceD = JabInjector.root.get<ServiceD>();
      expect(services.length, 4);
      expect(services, containsAllInOrder([serviceA, serviceB, serviceC, serviceD]));
    });

    test('get - onCreate callback reversed', () {
      final services = [];
      Jab.setOnCreate(services.add);

      Jab.provideForRoot(providers);

      expectAllABCD() {
        expect(services.length, 4);
        expect(services[0], isA<ServiceA>());
        expect(services[1], isA<ServiceB>());
        expect(services[2], isA<ServiceC>());
        expect(services[3], isA<ServiceD>());
      }

      JabInjector.root.get<ServiceD>();
      expectAllABCD();

      JabInjector.root.get<ServiceC>();
      expectAllABCD();

      JabInjector.root.get<ServiceB>();
      expectAllABCD();

      JabInjector.root.get<ServiceA>();
      expectAllABCD();
    });
  });
}


jab/test/jab_e2e_test.dart
import 'package:flutter/widgets.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:jab/jab.dart';
import 'package:jab_example_app/jab_example_app.dart';

void main() {
  // ignore: close_sinks
  final logger = Logger();

  group('Counter', () {
    setUp(() {
      Jab.provideForRoot([(_) => logger]);
    });

    tearDown(() {
      Jab.clearRoot();
    });

    testWidgets('Default', (WidgetTester tester) async {
      // Build our app and trigger a frame.
      await tester.pumpWidget(CounterApp());

      // -------------------------
      // - Open Default Counter
      // -------------------------

      // Go to Default Counter
      await tester.tap(find.text('Default Counter'));
      await tester.pumpAndSettle(Duration(seconds: 1));

      expect(logger.logs[0], "CounterApp.Jab.canCreate: CounterStore");
      expect(logger.logs[1], "CounterApp.Jab.onCreate: Instance of 'CounterStore'");

      // Verify that our counter starts at 0.00.
      expect(find.text('0.00'), findsOneWidget);
      expect(find.text('1.00'), findsNothing);

      // -------------------------
      // - Increment
      // -------------------------

      // Tap the 'Increment' button and trigger a frame.
      await tester.tap(find.byTooltip('Increment'));
      await tester.pump();

      expect(logger.logs[2], "CounterBloc.add: CounterEvent.increment");
      expect(logger.logs[3], "CounterStore.add: 1.0");

      // Verify that our counter has incremented to 1.00.
      expect(find.text('0.00'), findsNothing);
      expect(find.text('1.00'), findsOneWidget);

      // -------------------------
      // - Decrement
      // -------------------------

      // Tap the 'Decrement' button and trigger a frame.
      await tester.tap(find.byTooltip('Decrement'));
      await tester.pump();

      expect(logger.logs[4], "CounterBloc.add: CounterEvent.decrement");
      expect(logger.logs[5], "CounterStore.add: 0.0");

      // Verify that our counter has decremented to 0.00.
      expect(find.text('0.00'), findsOneWidget);
      expect(find.text('1.00'), findsNothing);

      // -------------------------
      // - Increment
      // -------------------------

      // Tap the 'Increment' button twice and trigger a frame.
      await tester.tap(find.byTooltip('Increment'));
      await tester.tap(find.byTooltip('Increment'));
      await tester.pump();

      expect(logger.logs[6], "CounterBloc.add: CounterEvent.increment");
      expect(logger.logs[7], "CounterStore.add: 1.0");
      expect(logger.logs[8], "CounterBloc.add: CounterEvent.increment");
      expect(logger.logs[9], "CounterStore.add: 2.0");

      // Verify that our counter has incremented to 2.00.
      expect(find.text('2.00'), findsOneWidget);
      expect(find.text('0.00'), findsNothing);

      // -------------------------
      // - Clear
      // -------------------------

      // Tap the 'Clear' button and trigger a frame.
      await tester.tap(find.byTooltip('Clear'));
      await tester.pump();

      expect(logger.logs[10], "CounterStore.clear");

      // Verify that our counter has cleared to 0.00.
      expect(find.text('2.00'), findsNothing);
      expect(find.text('0.00'), findsOneWidget);

      // -------------------------
      // - Go back
      // -------------------------

      // Go back
      await tester.tap(find.byTooltip('Back'));
      await tester.pumpAndSettle(Duration(seconds: 1));

      expect(logger.logs[11], "CounterBloc.close");

      // Pump a dummy Container
      await tester.pumpWidget(Container());

      expect(logger.logs[12], "CounterApp.Jab.onDispose: Instance of 'CounterStore'");
      expect(logger.logs[13], "CounterStore.close");
    });

    testWidgets('Golden', (WidgetTester tester) async {
      // Build our app and trigger a frame.
      await tester.pumpWidget(CounterApp());

      // -------------------------
      // - Open Golden Counter
      // -------------------------

      // Go to Default Counter
      await tester.tap(find.text('Golden Counter'));
      await tester.pumpAndSettle(Duration(seconds: 1));

      expect(logger.logs[0], "CounterApp.Jab.canCreate: CounterStore");
      expect(logger.logs[1], "CounterApp.Jab.onCreate: Instance of 'CounterStore'");

      // Verify that our counter starts at 0.00.
      expect(find.text('0.00'), findsOneWidget);
      expect(find.text('1.61'), findsNothing);

      // -------------------------
      // - Increment
      // -------------------------

      // Tap the 'Increment' button and trigger a frame.
      await tester.tap(find.byTooltip('Increment'));
      await tester.pump();

      expect(logger.logs[2], "GoldenCounterBloc.add: CounterEvent.increment");
      expect(logger.logs[3], "CounterStore.add: 1.618");

      // Verify that our counter has incremented to 1.00.
      expect(find.text('0.00'), findsNothing);
      expect(find.text('1.62'), findsOneWidget);

      // -------------------------
      // - Decrement
      // -------------------------

      // Tap the 'Decrement' button and trigger a frame.
      await tester.tap(find.byTooltip('Decrement'));
      await tester.pump();

      expect(logger.logs[4], "GoldenCounterBloc.add: CounterEvent.decrement");
      expect(logger.logs[5], "CounterStore.add: 1.0");

      // Verify that our counter has decremented to 0.00.
      expect(find.text('1.00'), findsOneWidget);
      expect(find.text('1.62'), findsNothing);

      // -------------------------
      // - Increment
      // -------------------------

      // Tap the 'Increment' button twice and trigger a frame.
      await tester.tap(find.byTooltip('Increment'));
      await tester.tap(find.byTooltip('Increment'));
      await tester.pump();

      expect(logger.logs[6], "GoldenCounterBloc.add: CounterEvent.increment");
      expect(logger.logs[7], "CounterStore.add: 1.618");
      expect(logger.logs[8], "GoldenCounterBloc.add: CounterEvent.increment");
      expect(logger.logs[9], "CounterStore.add: 2.6179240000000004");

      // Verify that our counter has incremented to 2.00.
      expect(find.text('2.62'), findsOneWidget);
      expect(find.text('1.00'), findsNothing);

      // -------------------------
      // - Clear
      // -------------------------

      // Tap the 'Clear' button and trigger a frame.
      await tester.tap(find.byTooltip('Clear'));
      await tester.pump();

      expect(logger.logs[10], "CounterStore.clear");

      // Verify that our counter has cleared to 0.00.
      expect(find.text('2.62'), findsNothing);
      expect(find.text('0.00'), findsOneWidget);

      // -------------------------
      // - Go back
      // -------------------------

      // Go back
      await tester.tap(find.byTooltip('Back'));
      await tester.pumpAndSettle(Duration(seconds: 1));

      expect(logger.logs[11], "GoldenCounterBloc.close");

      // Pump a dummy Container
      await tester.pumpWidget(Container());

      expect(logger.logs[12], "CounterApp.Jab.onDispose: Instance of 'CounterStore'");
      expect(logger.logs[13], "CounterStore.close");
    });
  });
}


lib/jab.dart
library jab;

import 'package:flutter/widgets.dart';

typedef JabProvider<T extends Object> = T Function(T Function<T extends Object>() get);

mixin JabTransientMixin {}

class JabTransientWrapper<T extends Object> {
  JabTransientWrapper(this.instance);

  final T instance;
}

extension JabTransientExtension<T extends Object> on T {
  JabTransientWrapper<T> get transient => JabTransientWrapper<T>(this);
}

/// A callback called before a Service instance is disposed.
///
/// Use this callback to initialize the Service.
typedef OnCreateCallback = void Function(Object service);

/// A callback called before a Service instance is disposed.
///
/// Use this callback to dispose any resources used by the Service.
typedef OnDisposeCallback = void Function(Object service);

mixin JabServiceMixin {
  void onCreate() {}

  void onDispose() {}
}

class JabInjector {
  JabInjector({this.onCreate, this.onDispose, this.parent});

  static final root = JabInjector();

  final OnCreateCallback? onCreate;

  final OnDisposeCallback? onDispose;

  final JabInjector? parent;

  final _providers = <JabProvider>{};

  final _services = <Type, Object>{};

  T get<T extends Object>() {
    if (_services[T] is T) {
      return _services[T] as T;
    }

    return _create<T>() ?? parent?.get<T>() ?? (throw _makeNotFoundException<T>());
  }

  T getTransient<T extends Object>() {
    return _create(transient: true) ?? parent?.getTransient<T>() ?? (throw _makeNotFoundException<T>());
  }

  void register<T extends Object>(JabProvider<T> provider) {
    _providers.add(provider);
  }

  void registerAll(Iterable<JabProvider> providers) {
    _providers.addAll(providers);
  }

  void dispose() {
    for (final service in _services.values) {
      if (service is JabServiceMixin) {
        service.onDispose();
      }
      onDispose?.call(service);
    }

    _providers.clear();
    _services.clear();
  }

  T? _create<T extends Object>({bool? transient}) {
    final provider = _providers.lastWhereOrNull((provider) {
      return provider is JabProvider<T> || provider is JabProvider<JabTransientWrapper<T>>;
    });

    if (provider == null) {
      return null;
    }

    final service = provider(get) as T;

    if (service is JabServiceMixin) {
      service.onCreate();
    }

    if (onCreate != null) {
      onCreate!(service);
    }

    if (service is JabTransientWrapper<T>) {
      return service.instance;
    } else if (service is JabTransientMixin || transient == true) {
      return service;
    } else {
      return _services[T] = service;
    }
  }

  Exception _makeNotFoundException<T>() {
    return Exception('A Service of type $T not found.');
  }
}

extension _IterableExtensions<E> on Iterable<E> {
  E? lastWhereOrNull(bool test(E element)) {
    try {
      return this.lastWhere(test);
    } catch (_) {
      return null;
    }
  }
}

class Jab extends StatefulWidget {
  const Jab({
    Key? key,
    this.injector,
    this.providers,
    this.onCreate,
    this.onDispose,
    required this.child,
  })   : assert(injector == null || (providers == null && onCreate == null && onDispose == null)),
        super(key: key);

  final JabInjector? injector;
  final Iterable<JabProvider> Function()? providers;

  final OnCreateCallback? onCreate;
  final OnDisposeCallback? onDispose;

  final Widget child;

  static JabInjector of(BuildContext context) {
    return _Jab.of(context);
  }

  static T get<T extends Object>(BuildContext context) {
    return of(context).get<T>();
  }

  static T getTransient<T extends Object>(BuildContext context) {
    return of(context).getTransient<T>();
  }

  @override
  JabState createState() => JabState();
}

class _Jab extends InheritedWidget {
  const _Jab({
    Key? key,
    required this.injector,
    required Widget child,
  }) : super(key: key, child: child);

  final JabInjector injector;

  static JabInjector of(BuildContext context) {
    final element = context.getElementForInheritedWidgetOfExactType<_Jab>();
    if (element != null) {
      return (element.widget as _Jab).injector;
    } else {
      return JabInjector.root;
    }
  }

  @override
  bool updateShouldNotify(_) {
    return false;
  }
}

class JabState extends State<Jab> {
  JabInjector get injector => _injector;
  late JabInjector _injector;

  @override
  void initState() {
    super.initState();
    _injector = widget.injector ?? _initInjecor();
    if (widget.providers != null) {
      _injector.registerAll(widget.providers!());
    }
  }

  @override
  void dispose() {
    _injector.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return _Jab(
      injector: _injector,
      child: widget.child,
    );
  }

  JabInjector _initInjecor() {
    return JabInjector(
      onCreate: widget.onCreate,
      onDispose: widget.onDispose,
      parent: Jab.of(context),
    );
  }
}

typedef JabServiceWidgetBuilder<T> = Widget Function(BuildContext context, T service);

class JabServiceBuilder<T extends Object> extends StatefulWidget {
  const JabServiceBuilder({
    Key? key,
    this.providers,
    required this.builder,
  }) : super(key: key);

  final Iterable<JabProvider> Function()? providers;
  final JabServiceWidgetBuilder<T> builder;

  @override
  _JabServiceBuilderState<T> createState() => _JabServiceBuilderState<T>();
}

class _JabServiceBuilderState<T extends Object> extends State<JabServiceBuilder<T>> {
  @override
  Widget build(BuildContext context) {
    if (widget.providers != null) {
      return Jab(
        providers: widget.providers ?? () => [],
        child: Builder(
          builder: (context) {
            return widget.builder(context, Jab.get<T>(context));
          },
        ),
      );
    } else {
      return widget.builder(context, Jab.get<T>(context));
    }
  }
}

typedef JabBlocWidgetBuilder<T> = Widget Function(BuildContext context, T bloc);

class JabBlocBuilder<T extends Object> extends StatefulWidget {
  const JabBlocBuilder({
    Key? key,
    this.providers,
    required this.builder,
  }) : super(key: key);

  final Iterable<JabProvider> Function()? providers;
  final JabBlocWidgetBuilder<T> builder;

  @override
  _JabBlocBuilderState<T> createState() => _JabBlocBuilderState<T>();
}

class _JabBlocBuilderState<T extends Object> extends State<JabBlocBuilder<T>> {
  @override
  Widget build(BuildContext context) {
    if (widget.providers != null) {
      return Jab(
        providers: widget.providers ?? () => [],
        child: Builder(
          builder: (context) {
            return widget.builder(context, Jab.getTransient<T>(context));
          },
        ),
      );
    } else {
      return widget.builder(context, Jab.getTransient<T>(context));
    }
  }
}

abstract class ViewStateBase {
  BuildContext get context;

  void initState();

  void dispose();
}

abstract class ViewState<T extends StatefulWidget> extends State<T> implements ViewStateBase {}

mixin BlocMixin<T extends Object> on ViewStateBase {
  T get bloc => _bloc;
  late T _bloc;

  @override
  void initState() {
    super.initState();
    _initBloc();
  }

  @override
  void dispose() {
    _disposeBloc();
    super.dispose();
  }

  void _initBloc() {
    _bloc = Jab.getTransient<T>(context);
  }

  void _disposeBloc() {
    if (bloc is Sink) {
      (bloc as Sink).close();
    }
  }
}

mixin ServiceMixin<T extends Object> on ViewStateBase {
  T get service => _service;
  late T _service;

  @override
  void initState() {
    super.initState();
    _initService();
  }

  void _initService() {
    _service = Jab.get<T>(context);
  }
}

abstract class StateWithBloc<T extends StatefulWidget, B extends Object> extends ViewState<T> with BlocMixin<B> {
  StateWithBloc({this.injector});

  final JabInjector? injector;

  @override
  void _initBloc() {
    if (injector != null) {
      _bloc = injector!.getTransient<B>();
    } else {
      super._initBloc();
    }
  }
}

abstract class StateWithService<T extends StatefulWidget, S extends Object> extends ViewState<T> with ServiceMixin<S> {
  StateWithService({this.injector});

  final JabInjector? injector;

  @override
  void _initService() {
    if (injector != null) {
      _service = injector!.get<S>();
    } else {
      super._initService();
    }
  }
}

class JabScope<S extends Object> extends StatefulWidget {
  const JabScope({
    Key? key,
    required this.child,
  }) : super(key: key);

  final Widget child;

  @override
  _JabScopeState<S> createState() => _JabScopeState<S>();
}

class _JabScopeState<S extends Object> extends State<JabScope<S>> {
  _JabScopeState({this.injector});

  final JabInjector? injector;

  S? get service => _service;
  S? _service;

  @override
  void initState() {
    super.initState();
    if (injector != null) {
      _service = injector!.getTransient<S>();
    } else {
      _service = Jab.get<S>(context);
    }
  }

  @override
  Widget build(BuildContext context) {
    return _JabScope(service: service, child: widget.child);
  }
}

class _JabScope<T extends Object> extends InheritedWidget {
  _JabScope({Key? key, required this.service, required Widget child}) : super(key: key, child: child);

  final T? service;

  T? maybeOf<T extends Object>(BuildContext context) {
    return context.getInheritedWidgetOfExactType<_JabScope<T>>()?.service;
  }

  @override
  bool updateShouldNotify(_JabScope<T> oldWidget) {
    return service != oldWidget.service;
  }
}

extension _InheritedWidgetBuildContextExtension on BuildContext {
  T? getInheritedWidgetOfExactType<T extends InheritedWidget>() {
    final widget = getElementForInheritedWidgetOfExactType<T>()?.widget;
    if (widget != null && widget is T) {
      return widget;
    }
  }
}


jab_example/lib/main.dart
import 'package:flutter/widgets.dart';
import 'package:jab/jab.dart';
import 'package:jab_example_app/jab_example_app.dart';

void main() {
  Jab.provideForRoot([
    (_) => Logger(),
  ]);
  runApp(CounterApp());
}

class JabProvider<T> {}

class FactoryProvider<T> extends JabProvider<T> {
  FactoryProvider(this.factory);

  final T Function() factory;
}

final providers = [
  FactoryProvider(() => Logger()),
  JabProvider<Logger>(),
];

final factories = [
  () => Logger(),
];

class NewJab {
  NewJab({
    this.factories,
    this.providers,
  });

  final Iterable<JabFactory> Function() factories;
  final Iterable<JabProvider> Function() providers;
}

final newJab = NewJab(
  factories: () => [
    (_) => Logger(),
  ],
  providers: () => [
    FactoryProvider(() => Logger()),
    JabProvider<Logger>(),
  ],
);

jab_example/lib/main_simple.dart
import 'package:flutter/foundation.dart';
import 'package:flutter/material.dart';
import 'package:jab/jab.dart';

void main() {
  runApp(CounterApp());
}

enum CounterEvent { increment, decrement, clear }

class CounterBloc extends ValueNotifier<int> implements Sink<CounterEvent> {
  CounterBloc() : super(0);

  @override
  void add(CounterEvent event) {
    switch (event) {
      case CounterEvent.increment:
        value = value + 1;
        break;
      case CounterEvent.decrement:
        value = value - 1;
        break;
      case CounterEvent.clear:
        value = 0;
        break;
    }
  }

  @override
  void close() {
    dispose();
  }
}

class CounterApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Jab(
      providers: () => [
        (_) => CounterBloc(),
      ],
      child: MaterialApp(
        title: 'Counter',
        home: CounterView(),
      ),
    );
  }
}

class CounterView extends StatefulWidget {
  @override
  _CounterViewState createState() => _CounterViewState();
}

class _CounterViewState extends ViewState<CounterView>
    with BlocMixin<CounterBloc> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text('Counter'),
      ),
      body: ValueListenableBuilder<int>(
        valueListenable: bloc,
        builder: (BuildContext context, int value, _) {
          return Center(
            child: Text(
              value.toString(),
              style: TextStyle(fontSize: 48),
            ),
          );
        },
      ),
      floatingActionButton: Column(
        mainAxisSize: MainAxisSize.min,
        children: <Widget>[
          IncrementButton(
            onPressed: () {
              bloc.add(CounterEvent.increment);
            },
          ),
          SizedBox(height: 16),
          DecrementButton(
            onPressed: () {
              bloc.add(CounterEvent.decrement);
            },
          ),
          SizedBox(height: 16),
          ClearButton(
            onPressed: () {
              bloc.add(CounterEvent.clear);
            },
          )
        ],
      ),
    );
  }
}

class IncrementButton extends StatelessWidget {
  const IncrementButton({Key key, this.onPressed}) : super(key: key);

  final VoidCallback onPressed;

  @override
  Widget build(BuildContext context) {
    return FloatingActionButton(
      heroTag: 'Increment',
      tooltip: 'Increment',
      child: Icon(Icons.add),
      onPressed: onPressed,
    );
  }
}

class DecrementButton extends StatelessWidget {
  const DecrementButton({Key key, this.onPressed}) : super(key: key);

  final VoidCallback onPressed;

  @override
  Widget build(BuildContext context) {
    return FloatingActionButton(
      heroTag: 'Decrement',
      tooltip: 'Decrement',
      child: Icon(Icons.remove),
      onPressed: onPressed,
    );
  }
}

class ClearButton extends StatelessWidget {
  const ClearButton({Key key, this.onPressed}) : super(key: key);

  final VoidCallback onPressed;

  @override
  Widget build(BuildContext context) {
    return FloatingActionButton(
      heroTag: 'Clear',
      tooltip: 'Clear',
      child: Icon(Icons.delete),
      onPressed: onPressed,
    );
  }
}


jab_example_app/lib/jab_example_app.dart
library jab_example_app;

import 'dart:async';

import 'package:flutter/material.dart';
import 'package:jab/jab.dart';
import 'package:rxdart/rxdart.dart';

class Logger implements Sink<String> {
  List<String> get logs => _logs;
  final _logs = List<String>();

  @override
  void add(String data) {
    print('[Logger] $data');
    _logs.add(data);
  }

  @override
  void close() {
    _logs.clear();
  }
}

class CounterStore extends Stream<double> implements Sink<double> {
  CounterStore(this.logger);

  final _controller = BehaviorSubject<double>.seeded(0.0);

  final Logger logger;

  @override
  void add(double data) {
    logger.add('CounterStore.add: $data');
    _controller.add(data);
  }

  @override
  void close() {
    logger.add('CounterStore.close');
    _controller.close();
  }

  void clear() {
    logger.add('CounterStore.clear');
    _controller.add(0);
  }

  @override
  StreamSubscription<double> listen(
    void Function(double event) onData, {
    Function onError,
    void Function() onDone,
    bool cancelOnError,
  }) {
    return _controller.listen(
      onData,
      onError: onError,
      onDone: onDone,
      cancelOnError: cancelOnError,
    );
  }
}

enum CounterEvent { increment, decrement }

class CounterBloc implements Sink<CounterEvent> {
  final CounterStore _counterStore;
  final Logger _logger;

  StreamSubscription<double> _subscription;
  double _currentValue;

  CounterBloc(this._counterStore, this._logger) {
    _subscription = _counterStore.listen((value) => _currentValue = value);
  }

  @override
  void add(CounterEvent event) {
    _logger.add('CounterBloc.add: $event');
    switch (event) {
      case CounterEvent.increment:
        _counterStore.add(_currentValue + 1);
        break;
      case CounterEvent.decrement:
        _counterStore.add(_currentValue - 1);
        break;
    }
  }

  @override
  void close() {
    _subscription.cancel();
    _logger.add('CounterBloc.close');
  }
}

class GoldenCounterBloc extends CounterBloc {
  GoldenCounterBloc(CounterStore counterStore, Logger logger)
      : super(counterStore, logger);

  @override
  void add(CounterEvent event) {
    _logger.add('GoldenCounterBloc.add: $event');
    switch (event) {
      case CounterEvent.increment:
        if (_currentValue == 0) {
          _currentValue = 1;
        }
        _counterStore.add(_currentValue * 1.618);
        break;
      case CounterEvent.decrement:
        _counterStore.add(_currentValue / 1.618);
        break;
    }
  }

  @override
  void close() {
    _subscription.cancel();
    _logger.add('GoldenCounterBloc.close');
  }
}

class CounterApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Jab(
      providers: () => [
        (jab) => CounterStore(jab()),
        (jab) => CounterBloc(jab(), jab()),
      ],
      canCreate: (type) {
        Jab.get<Logger>(context).add('CounterApp.Jab.canCreate: $type');
        return true;
      },
      onCreate: (service) {
        Jab.get<Logger>(context).add('CounterApp.Jab.onCreate: $service');
      },
      onDispose: (service) {
        JabInjector.root
            .get<Logger>()
            .add('CounterApp.Jab.onDispose: $service');
      },
      child: MaterialApp(
        title: 'Jab Examples',
        routes: {
          '/': (_) => CounterAppHome(),
          '/default': (_) => CounterView(),
          '/golden': (_) => GoldenCounterView(),
        },
      ),
    );
  }
}

class GoldenCounterView extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Jab(
      providers: () => [
        (jab) => GoldenCounterBloc(jab(), jab()),
      ],
      child: CounterView(),
    );
  }
}

class CounterAppHome extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text('Jab Examples'),
      ),
      body: ListView(
        children: <Widget>[
          ListTile(
            title: Text('Default Counter'),
            trailing: Icon(Icons.arrow_forward_ios),
            onTap: () {
              Navigator.pushNamed(context, '/default');
            },
          ),
          ListTile(
            title: Text('Golden Counter'),
            trailing: Icon(Icons.arrow_forward_ios),
            onTap: () {
              Navigator.pushNamed(context, '/golden');
            },
          )
        ],
      ),
    );
  }
}

class CounterView extends StatefulWidget {
  @override
  _CounterViewState createState() => _CounterViewState();
}

class _CounterViewState extends ViewState<CounterView>
    with BlocMixin<CounterBloc> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text('Jab Counter'),
      ),
      body: StreamBuilder<double>(
        stream: Jab.get<CounterStore>(context),
        builder: (context, snapshot) {
          return Center(
            child: Text(
              snapshot.data?.toStringAsFixed(2) ?? 'N/A',
              style: TextStyle(fontSize: 48),
              textAlign: TextAlign.center,
            ),
          );
        },
      ),
      floatingActionButton: Column(
        mainAxisSize: MainAxisSize.min,
        children: <Widget>[
          IncrementButton(
            onPressed: () {
              bloc.add(CounterEvent.increment);
            },
          ),
          SizedBox(height: 16),
          DecrementButton(
            onPressed: () {
              bloc.add(CounterEvent.decrement);
            },
          ),
          SizedBox(height: 16),
          ClearButton(
            onPressed: () {
              Jab.get<CounterStore>(context).clear();
            },
          ),
        ],
      ),
    );
  }
}

class IncrementButton extends StatelessWidget {
  const IncrementButton({Key key, this.onPressed}) : super(key: key);

  final VoidCallback onPressed;

  @override
  Widget build(BuildContext context) {
    return FloatingActionButton(
      heroTag: 'Increment',
      tooltip: 'Increment',
      child: Icon(Icons.arrow_upward),
      onPressed: onPressed,
    );
  }
}

class DecrementButton extends StatelessWidget {
  const DecrementButton({Key key, this.onPressed}) : super(key: key);

  final VoidCallback onPressed;

  @override
  Widget build(BuildContext context) {
    return FloatingActionButton(
      heroTag: 'Decrement',
      tooltip: 'Decrement',
      child: Icon(Icons.arrow_downward),
      onPressed: onPressed,
    );
  }
}

class ClearButton extends StatelessWidget {
  const ClearButton({Key key, this.onPressed}) : super(key: key);

  final VoidCallback onPressed;

  @override
  Widget build(BuildContext context) {
    return FloatingActionButton(
      heroTag: 'Clear',
      tooltip: 'Clear',
      child: Icon(Icons.delete),
      onPressed: onPressed,
    );
  }
}
